crates/matrix-sdk-ui/src/timeline/pagination.rs

```diff
@@ -12,9 +12,108 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::{fmt, ops::ControlFlow};
+use std::{fmt, ops::ControlFlow, pin::pin, sync::Arc, time::Duration};
+
+use matrix_sdk::{room::MessagesOptions, Result};
+use matrix_sdk_base::timeout::timeout;
+use ruma::assign;
+use tracing::{debug, error, info, warn};
+
+use super::{inner::HandleBackPaginatedEventsError, Timeline};
+
+impl Timeline {
+    pub(super) async fn paginate_backwards_impl(
+        &self,
+        mut options: PaginationOptions<'_>,
+    ) -> Result<ControlFlow<BackPaginationStatus>> {
+        let mut from = match self.inner.back_pagination_token().await {
+            None if options.wait_for_token => {
+                let notified = pin!(self.sync_response_notify.notified());
+                match timeout(notified, Duration::from_secs(3)).await {
+                    Ok(()) => match self.inner.back_pagination_token().await {
+                        Some(token) => Some(token),
+                        None => {
+                            warn!(
+                                "Sync response without prev_batch received, \
+                                 start of timeline reached?"
+                            );
+                            return Ok(ControlFlow::Break(BackPaginationStatus::Idle));
+                        }
+                    },
+                    Err(_) => {
+                        debug!("Waiting for prev_batch token timed out after 3s");
+                        None
+                    }
+                }
+            }
+            token => token,
+        };
+
+        self.back_pagination_status.set_if_not_eq(BackPaginationStatus::Paginating);
```

Comment:
I find it a bit weird that we're setting the back_pagination_status both here and in the caller. Instead, could we remove the status data from the `Break` type, and always set the status in this function? (except for errors, of course)

Reply:
By declaring the return type to be `BackPaginationStatus`, we can't accidentally introduce new early returns that don't set the status.